# Data Exploration

EDA (Exploratory Data Analysis) ini merupakan bagian dimana kita akan melakukan eksplorasi data dengan tujuan untuk memahami data dan bahkan menjawab pertanyaan dengan data. Pada bagian data exploration ini kita akan mengaplikasikan teknik statistik yang sudah dipelajari sebelumnya.

Contoh explore data (exploring data):
- cara import data
- cara mendapatkan insight dari data

menjawab pertanyaan dengan data (answering question):
- mencari fakta dari data
- menggali informasi dari data

bagaimana cara explore data:
- time to practice

In [2]:
import pandas as pd
import numpy as np

#### Exploring Data

sekarang kita akan import data dari website. Data ini adalah data tentang penggunaan alkohol di seluruh dunia.

In [3]:
df = pd.read_csv('http://bit.ly/drinksbycountry')
df.head()

country  beer_servings  spirit_servings  wine_servings  \
0  Afghanistan              0                0              0   
1      Albania             89              132             54   
2      Algeria             25                0             14   
3      Andorra            245              138            312   
4       Angola            217               57             45   

   total_litres_of_pure_alcohol continent  
0                           0.0      Asia  
1                           4.9    Europe  
2                           0.7    Africa  
3                          12.4    Europe  
4                           5.9    Africa

kita akan mencoba mendapatkan insight dengan fungsi yang sederhana

In [4]:
df.describe()

beer_servings  spirit_servings  wine_servings  \
count     193.000000       193.000000     193.000000   
mean      106.160622        80.994819      49.450777   
std       101.143103        88.284312      79.697598   
min         0.000000         0.000000       0.000000   
25%        20.000000         4.000000       1.000000   
50%        76.000000        56.000000       8.000000   
75%       188.000000       128.000000      59.000000   
max       376.000000       438.000000     370.000000   

       total_litres_of_pure_alcohol  
count                    193.000000  
mean                       4.717098  
std                        3.773298  
min                        0.000000  
25%                        1.300000  
50%                        4.200000  
75%                        7.200000  
max                       14.400000

describe() digunakan untuk menjabarkan data yang kita miliki dalam bentuk statistik namun kolom yang dapat didescribe hanyalah kolom yang berisi data angka.

Dari data di atas kita bisa mengetahui mayoritas negara mengonsumsi beer dengan melihat hubungan antara mean dan std (**mean - std** hingga **mean + std**). Kita juga dapat mengetahui bahwa ada negara yang tidak mengonsumsi alkohol sama sekali dari nilai minimum (0). Kita juga bisa memprediksi outlier, coba lihat pada bagian **wine_servings**, nilai mediannya saja 8, meannya hanya 49 dan Q3 nya hanya 59 tetapi nilai maksimumnya mencapai 370 yang menunjukkan bahwa kemungkinan besar ada outlier.


apakah data string tidak bisa dilihat data statistiknya? **bisa duonk**

gunakan describe(include='all')

In [5]:
df.describe(include='all')

country  beer_servings  spirit_servings  wine_servings  \
count      193     193.000000       193.000000     193.000000   
unique     193            NaN              NaN            NaN   
top       Togo            NaN              NaN            NaN   
freq         1            NaN              NaN            NaN   
mean       NaN     106.160622        80.994819      49.450777   
std        NaN     101.143103        88.284312      79.697598   
min        NaN       0.000000         0.000000       0.000000   
25%        NaN      20.000000         4.000000       1.000000   
50%        NaN      76.000000        56.000000       8.000000   
75%        NaN     188.000000       128.000000      59.000000   
max        NaN     376.000000       438.000000     370.000000   

        total_litres_of_pure_alcohol continent  
count                     193.000000       193  
unique                           NaN         6  
top                              NaN    Africa  
freq                             NaN        53  
mean                        4.717098       NaN  
std                         3.773298       NaN  
min                         0.000000       NaN  
25%                         1.300000       NaN  
50%                         4.200000       NaN  
75%                         7.200000       NaN  
max                        14.400000       NaN

jika kita lihat kolom country, **unique** adalah jumlah data unik, **top** adalah data yang paling sering keluar dan berhubung semua negara unik sehingga semua negara bisa menjadi **top** dengan **freq** nya 1. Begitu pula dengan continent, namun **top** nya adalah Africa dengan **freq** nya berjumlah 53 artinya ada 53 data berisi Africa.

Bagaimana jika kita ingin mengetahui info kolom dalam data?

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 6 columns):
country                         193 non-null object
beer_servings                   193 non-null int64
spirit_servings                 193 non-null int64
wine_servings                   193 non-null int64
total_litres_of_pure_alcohol    193 non-null float64
continent                       193 non-null object
dtypes: float64(1), int64(3), object(2)
memory usage: 9.1+ KB


info di atas menunjukkan banyak informasi seperti type data, bahkan kolom yang memiliki nilai kosong. Kebetulan dari data di atas semuanya **non-null** artinya tidak ada nilai kosong. Kita juga bisa mengetahui memory usage dari data tersebut.

#### Answering the Question

- Ada berapa total country dari tiap continent dari dataset?

In [7]:
df['continent'].value_counts() #yang ini cukup sih sebenernya

Africa           53
Europe           45
Asia             44
North America    23
Oceania          16
South America    12
Name: continent, dtype: int64

In [8]:
df['continent'].value_counts().sort_values()

South America    12
Oceania          16
North America    23
Asia             44
Europe           45
Africa           53
Name: continent, dtype: int64

In [9]:
df['continent'].value_counts().sort_values(ascending=False)

Africa           53
Europe           45
Asia             44
North America    23
Oceania          16
South America    12
Name: continent, dtype: int64

- Berapa negara di Asia yang tidak mengonsumsi beer?

Kita akan melakukan filtering, pertama kita akan melihat negara negara di Asia

In [10]:
df[df['continent'] == 'Asia']

country  beer_servings  spirit_servings  wine_servings  \
0             Afghanistan              0                0              0   
12                Bahrain             42               63              7   
13             Bangladesh              0                0              0   
19                 Bhutan             23                0              0   
24                 Brunei             31                2              1   
30               Cambodia             57               65              1   
36                  China             79              192              8   
46            North Korea              0                0              0   
77                  India              9              114              0   
78              Indonesia              5                1              0   
79                   Iran              0                0              0   
80                   Iraq              9                3              0   
82                 Israel             63               69              9   
85                  Japan             77              202             16   
86                 Jordan              6               21              1   
87             Kazakhstan            124              246             12   
90                 Kuwait              0                0              0   
91             Kyrgyzstan             31               97              6   
92                   Laos             62                0            123   
94                Lebanon             20               55             31   
102              Malaysia             13                4              0   
103              Maldives              0                0              0   
112              Mongolia             77              189              8   
116               Myanmar              5                1              0   
119                 Nepal              5                6              0   
127                  Oman             22               16              1   
128              Pakistan              0                0              0   
134           Philippines             71              186              1   
137                 Qatar              1               42              7   
138           South Korea            140               16              9   
141    Russian Federation            247              326             73   
149          Saudi Arabia              0                5              0   
154             Singapore             60               12             11   
161             Sri Lanka             16              104              0   
167                 Syria              5               35             16   
168            Tajikistan              2               15              0   
169              Thailand             99              258              1   
171           Timor-Leste              1                1              4   
176                Turkey             51               22              7   
177          Turkmenistan             19               71             32   
181  United Arab Emirates             16              135              5   
186            Uzbekistan             25              101              8   
189               Vietnam            111                2              1   
190                 Yemen              6                0              0   

     total_litres_of_pure_alcohol continent  
0                             0.0      Asia  
12                            2.0      Asia  
13                            0.0      Asia  
19                            0.4      Asia  
24                            0.6      Asia  
30                            2.2      Asia  
36                            5.0      Asia  
46                            0.0      Asia  
77                            2.2      Asia  
78                            0.1      Asia  
79                            0.0      Asia  
80                            0.2      Asia  

sip, sekarang kita akan melihat negara Asia yang **beer_servings** nya samadengan **0**

In [11]:
df[(df['continent'] == 'Asia') & (df['beer_servings'] == 0)]

country  beer_servings  spirit_servings  wine_servings  \
0     Afghanistan              0                0              0   
13     Bangladesh              0                0              0   
46    North Korea              0                0              0   
79           Iran              0                0              0   
90         Kuwait              0                0              0   
103      Maldives              0                0              0   
128      Pakistan              0                0              0   
149  Saudi Arabia              0                5              0   

     total_litres_of_pure_alcohol continent  
0                             0.0      Asia  
13                            0.0      Asia  
46                            0.0      Asia  
79                            0.0      Asia  
90                            0.0      Asia  
103                           0.0      Asia  
128                           0.0      Asia  
149                           0.1      Asia

data di atas adalah negara-negara di Asia yang tidak mengonsumsi beer, sekarang bagaimana jika kita ingin menghitung jumlahnya?

In [12]:
len(df[(df['continent'] == 'Asia') & (df['beer_servings'] == 0)])

8

- negara di Eropa yang mengonsumsi wine dengan intesitas lebih dari 100

sama seperti sebelumnya tapi kita akan mencari negara di Eropa yang **wine_servings**nya lebih dari 100

In [13]:
df[(df['continent'] == 'Europe') & (df['wine_servings'] > 100)]

country  beer_servings  spirit_servings  wine_servings  \
3           Andorra            245              138            312   
9           Austria            279               75            191   
16          Belgium            295               84            212   
42          Croatia            230               87            254   
44           Cyprus            192              154            113   
45   Czech Republic            361              170            134   
48          Denmark            224               81            278   
61           France            127              151            370   
64          Georgia             52              100            149   
65          Germany            346              117            175   
67           Greece            133              112            218   
75          Hungary            234              215            185   
81          Ireland            313              118            165   
83            Italy             85               42            237   
99       Luxembourg            236              133            271   
105           Malta            149              100            120   
113      Montenegro             31              114            128   
120     Netherlands            251               88            190   
126          Norway            169               71            129   
136        Portugal            194               67            339   
140         Romania            297              122            167   
151          Serbia            283              131            127   
155        Slovakia            196              293            116   
156        Slovenia            270               51            276   
160           Spain            284              157            112   
165          Sweden            152               60            186   
166     Switzerland            185              100            280   
182  United Kingdom            219              126            195   

     total_litres_of_pure_alcohol continent  
3                            12.4    Europe  
9                             9.7    Europe  
16                           10.5    Europe  
42                           10.2    Europe  
44                            8.2    Europe  
45                           11.8    Europe  
48                           10.4    Europe  
61                           11.8    Europe  
64                            5.4    Europe  
65                           11.3    Europe  
67                            8.3    Europe  
75                           11.3    Europe  
81                           11.4    Europe  
83                            6.5    Europe  
99                           11.4    Europe  
105                           6.6    Europe  
113                           4.9    Europe  
120                           9.4    Europe  
126                           6.7    Europe  
136                          11.0    Europe  
140                          10.4    Europe  
151                           9.6    Europe  
155                          11.4    Europe  
156                          10.6    Europe  
160                          10.0    Europe  
165                           7.2    Europe  
166                          10.2    Europe  
182                          10.4    Europe

In [14]:
len(df[(df['continent'] == 'Europe') & (df['wine_servings'] > 100)])

28

- negara apa yang paling banyak mengonsumsi spirit di benua Afrika?

kita lihat dulu negara-negara di Afrika

In [15]:
df[df['continent'] == 'Africa']

country  beer_servings  spirit_servings  wine_servings  \
2                     Algeria             25                0             14   
4                      Angola            217               57             45   
18                      Benin             34                4             13   
22                   Botswana            173               35             35   
26               Burkina Faso             25                7              7   
27                    Burundi             88                0              0   
28              Cote d'Ivoire             37                1              7   
29                 Cabo Verde            144               56             16   
31                   Cameroon            147                1              4   
33   Central African Republic             17                2              1   
34                       Chad             15                1              1   
38                    Comoros              1                3              1   
39                      Congo             76                1              9   
47                   DR Congo             32                3              1   
49                   Djibouti             15               44              3   
53                      Egypt              6                4              1   
55          Equatorial Guinea             92                0            233   
56                    Eritrea             18                0              0   
58                   Ethiopia             20                3              0   
62                      Gabon            347               98             59   
63                     Gambia              8                0              1   
66                      Ghana             31                3             10   
70                     Guinea              9                0              2   
71              Guinea-Bissau             28               31             21   
88                      Kenya             58               22              2   
95                    Lesotho             82               29              0   
96                    Liberia             19              152              2   
97                      Libya              0                0              0   
100                Madagascar             26               15              4   
101                    Malawi              8               11              1   
104                      Mali              5                1              1   
107                Mauritania              0                0              0   
108                 Mauritius             98               31             18   
114                   Morocco             12                6             10   
115                Mozambique             47               18              5   
117                   Namibia            376                3              1   
123                     Niger              3                2              1   
124                   Nigeria             42                5              2   
142                    Rwanda             43                2              0   
148       Sao Tome & Principe             56               38            140   
150                   Senegal              9                1              7   
152                Seychelles            157               25             51   
153              Sierra Leone             25                3              2   
158                   Somalia              0                0              0   
159              South Africa            225               76             81   
162                     Sudan              8               13              0   
164                 Swaziland             90                2              2   
172                      Togo             36                2             19   
175                   Tunisia             51                3             20   
179                   

sekarang kita sorting secara **descending** berdasarkan **spirit_servings**. Karena kita hanya ingin tahu yang paling banyak mengonsumsi spirit jadi kita cukup ambil data teratasnya saja

In [17]:
df[(df['continent'] == 'Africa')].sort_values(['spirit_servings'], ascending=False).head(1)

country  beer_servings  spirit_servings  wine_servings  \
96  Liberia             19              152              2   

    total_litres_of_pure_alcohol continent  
96                           3.1    Africa

sip, sekarang kita tau bahwa negara di benua Afrika yang mengonsumsi spirit terbanyak adalah Liberia. Haram ya jangan ikutan!

- berapa rata-rata konsumsi alkohol dari tiap benua?

langsung aja lah ya biar cepet, ez ini mah

In [21]:
df.groupby(['continent']).mean().sort_values(['total_litres_of_pure_alcohol'], ascending=False)

beer_servings  spirit_servings  wine_servings  \
continent                                                      
Europe            193.777778       132.555556     142.222222   
South America     175.083333       114.750000      62.416667   
North America     145.434783       165.739130      24.521739   
Oceania            89.687500        58.437500      35.625000   
Africa             61.471698        16.339623      16.264151   
Asia               37.045455        60.840909       9.068182   

               total_litres_of_pure_alcohol  
continent                                    
Europe                             8.617778  
South America                      6.308333  
North America                      5.995652  
Oceania                            3.381250  
Africa                             3.007547  
Asia                               2.170455

nah eta